subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java

```diff
@@ -0,0 +1,122 @@
+    private static final String TEST_BUNDLE_SYMBOLIC_NAME = "testBundle";
+
+    private static final long STOP_TIMEOUT_MS = 10000;
+
+    private static Path frameworkStorage;
+    private static Framework framework;
+    private static Bundle testBundle;
+
+    @BeforeClass
+    public static void setUp() throws IOException, BundleException {
+        frameworkStorage = Files.createTempDirectory(STORAGE_TEMPDIR_NAME);
+        Map<String, String> configuration = new HashMap<>();
+        configuration.put(Constants.FRAMEWORK_STORAGE, frameworkStorage.toString());
+        configuration.put(Constants.FRAMEWORK_SYSTEMPACKAGES_EXTRA, String.join(",", EXTRA_SYSTEMPACKAGES));
+        framework = frameworkFactory.newFramework(configuration);
+        framework.init();
+        BundleContext bundleContext = framework.getBundleContext();
+        for (Path dependencyPath : TEST_RUNTIME_BUNDLES) {
+            Bundle installedBundle;
+            try {
+                installedBundle = bundleContext.installBundle(dependencyPath.toUri().toString());
+            } catch (BundleException e) {
+                throw new IllegalStateException("Failed to install bundle: " + dependencyPath.getFileName(), e);
+            }
+            if (TEST_BUNDLE_SYMBOLIC_NAME.equals(installedBundle.getSymbolicName())) {
+                testBundle = installedBundle;
+            }
+        }
+        if (testBundle == null) {
+            fail("Test bundle not found.");
+        }
+        framework.start();
+    }
+
+    @AfterClass
+    public static void tearDown() throws IOException, BundleException, InterruptedException {
+        try {
+            if (framework != null) {
+                framework.stop();
+                framework.waitForStop(STOP_TIMEOUT_MS);
+            }
+        } finally {
+            if (frameworkStorage != null) {
+                deleteRecursively(frameworkStorage);
+            }
+        }
+    }
+
+    @Test
+    public void testSimpleMock() throws BundleException, InterruptedException, ClassNotFoundException, IllegalAccessException, InstantiationException, IOException {
+        createTestRunnable("SimpleMockTest").run();
+    }
+
+    @Test
+    public void testMockNonPublicClassFails() throws BundleException, InterruptedException, ClassNotFoundException, IllegalAccessException, InstantiationException, IOException {
+        createTestRunnable("MockNonPublicClassFailsTest").run();
+    }
+
+    @Test
+    public void testMockClassInOtherBundle() throws BundleException, InterruptedException, ClassNotFoundException, IllegalAccessException, InstantiationException, IOException {
+        createTestRunnable("MockClassInOtherBundleTest").run();
+    }
+
+    private Runnable createTestRunnable(String className) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
+        return (Runnable) testBundle.loadClass("org.mockito.osgitest.testbundle." + className).newInstance();
```

Comment:
Can we use the JUnit suite API to run these tests? For an example, see https://github.com/mockito/mockito/blob/release/3.x/src/test/java/org/mockitousage/junitrunner/StrictRunnerTest.java and the `JUnitCore` API.

Reply:
I did use the suite API to make the tests run directly in the Gradle runner, which I think is better than using an explicit `JUnitCore` runner in this situation.